### ISSUE TITLE
[Wallet] Possible race condition while stopping running synchronization upon unlock wallet

### ISSUE THREAD

**Issue by mariocao** [witnet/witnet-rust#1849]
If a new wallet is unlocked and there was an on-going synchronization process, a `stop_syncing` flag is set to true in order to signal the wallet to stop synchronizing:

```rust
    pub fn unlock_wallet(
        &mut self,
        wallet_id: String,
        password: types::Password,
    ) -> ResponseActFuture<types::UnlockedWallet> {
        // If a synchronization from a previous session is still running, set `stop_syncing` to
        // `true` so as to signal that it must stop as soon as possible
        if let Some(wallet) = self.state.get_current_wallet_session(wallet_id.clone()) {
            wallet.set_stop_syncing().expect("Lock error")
        }
```

However, this flag is read in the `sync_inner` function in `worker/methods.rs`:
```rust
     loop {
            // ...

            // For each of the blocks we have been informed about, ask a Witnet node for its contents
            for ChainEntry(_epoch, id) in block_chain {
                let get_block_future = self.get_block(id.clone());
                let (block, confirmed) = futures::executor::block_on(get_block_future)?;

                // Wrap block into an atomic reference count for the sake of avoiding expensive clones
                let block_arc = Arc::new(block);

                // Process each block and update latest beacon
                self.handle_block(
                    block_arc.clone(),
                    confirmed,
                    wallet.clone(),
                    DynamicSink::default(),
                )?;
                latest_beacon = block_arc.block_header.beacon;
            }

            // ...

            // Keep asking for new batches of blocks until we get less than expected, which signals
            // that there are no more blocks to process.
            if batch_size < i128::from(limit)
                || wallet.lock_and_read_state(|state| state.stop_syncing)?
            {
                break;
            }
    }
```

It seems that there could be some race condition with the main consequence that some block batches may be processed twice by the same wallet. However, this should be further investigated.

**Comment by mariocao** [comment]
@lrubiorod and I tried to check if there was a race condition regarding the aforementioned code. We were not able to make 2 synchronization processes from the same wallet index the same block (`handle_block` may return an `Err`).

A possible improvement is to add a break condition inside the `for` loop where blocks are being handled:

```rust
if wallet.lock_and_read_state(|state| state.stop_syncing)? {
  break;
}
```

However, we decided not to include this improvement because we were not able to intercept an Error from the `handle_block` function.
